### ISSUE TITLE
Handle ECN bits properly

### ISSUE THREAD

**Issue by thomaseizinger** [CloudFire-LLC/cloudfire#3758]
The `quinn-udp` socket gives us access to the ECN bits set on the IP datagram on the wire. We should explore if it makes sense to capture those and copy them to the IP packet sent out via the TUN device. Vice-versa, we should set the ECN bits upon sending correctly.

This might reduce packet drops and thus improve reliability of the connection.

**Comment by thomaseizinger** [comment]
This might not be as difficult as I thought it would be:

- Firezone should always set ECN bits on the UDP datagrams it generates
- If we detect congestion on incoming packets, we need to look at the bits of the encapsulated IP packet:
   - If it has ECN bits set, meaning the user space application supports ECN, also set the bit to signal congestion
   - If it doesn't have them set, drop the packet.

This should allow for a smoother adjustment transfer rates because middleboxes won't drop our UDP traffic once we start being too fast but instead flip the bit which allows the user space application to slow down without experiencing packet loss.

**Comment by thomaseizinger** [comment]
Potentially relevant: https://www.rfc-editor.org/rfc/rfc6040.html

**Comment by thomaseizinger** [comment]
> * If it doesn't have them set, drop the packet.

Not sure this is going to super useful. If the upper layer _doesn't_ support ECN bits, then it doesn't make much of a difference whether we are dropping the packet or some intermediary along the path drops it (instead of setting the congestion bit).

The main use we probably can get out of ECN is mirroring whatever bit the application set. If it wants to signal support for ECN, we should mirror that on our UDP packets to avoid them being dropped along the way. Upon decapsulation, we should check whether our UDP packet got flagged as congested and mirror that on the decapsulated packet to inform the receiving end.

**Comment by jamilbk** [comment]
Is this still relevant / what priority should we give it (what user impact could it have?)

**Comment by thomaseizinger** [comment]
> Is this still relevant / what priority should we give it (what user impact could it have?)

It is a performance optimisation that allows ECN-aware applications / congestion controllers to perform a smoother adjustment of the traffic they are sending instead of relying on measure packetloss.

Likely a very easy one to implement because we just need to mirror what is set on the inner packet. I would put it high up in priority once somebody works on performance improvement next.

**Comment by thomaseizinger** [comment]
The official WG implementation also does this (and references RFC6040 for it). See https://www.wireguard.com/protocol/#diffserv-considerations.

**Comment by thomaseizinger** [comment]
#8511 implements copying the ECN bits back and forth between the two IP layers in `connlib`. According to my research, this might not be enough though because the default in Linux is still to not actively announce ECN capability but accept it when a remote party opens a TCP connection with ECN bits set.

In our case, the Gateway will open all connections to resources and therefore, we won't actually see any change with this from just #8511.

Flipping this setting to always use ECN is just one `sysctl` command away, the question is do we want to include that in our Gateway start up scripts?

To always perform ECN, we need to run: `sysctl -w net.ipv4.tcp_ecn=1` (the default is 2).

**Comment by jamilbk** [comment]
Hmmm, is there any downside to setting it? If it won't cause an issue then yeah we could add it to our startup scripts. Would it make more sense to call this from within the Rust code though so existing installs benefit from it?

**Comment by thomaseizinger** [comment]
I'll have to read up more on the risks.

**Comment by thomaseizinger** [comment]
> In our case, the Gateway will open all connections to resources and therefore, we won't actually see any change with this from just [#8511](https://github.com/firezone/firezone/pull/8511).

This is wrong, the client opens the TCP connection, we just tunnel it to the Gateway. So whether or not ECN bits are set depends on what the client network stack does.
